src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -222,51 +223,242 @@ private bool ScanDesignation(bool permitTuple)
+            switch (this.CurrentToken.Kind)
+            {
+                case SyntaxKind.CommaToken:
+                case SyntaxKind.SemicolonToken:
+                case SyntaxKind.CloseBraceToken:
+                case SyntaxKind.CloseParenToken:
+                case SyntaxKind.CloseBracketToken:
+                case SyntaxKind.EqualsGreaterThanToken:
+                    return this.ParseIdentifierName(ErrorCode.ERR_MissingArgument);
+            }
 
-            // If it is a nameof, skip the 'if' and parse as an expression. 
-            if ((SyntaxFacts.IsPredefinedType(tk) || tk == SyntaxKind.IdentifierToken) &&
-                  this.CurrentToken.ContextualKind != SyntaxKind.NameOfKeyword)
+            var resetPoint = this.GetResetPoint();
+            try
             {
-                var resetPoint = this.GetResetPoint();
-                try
+                TypeSyntax type = null;
+                var tk = this.CurrentToken.Kind;
+                if ((SyntaxFacts.IsPredefinedType(tk) || tk == SyntaxKind.IdentifierToken) &&
+                      // If it is a nameof, skip the 'if' and parse as an expression. 
+                      (this.CurrentToken.ContextualKind != SyntaxKind.NameOfKeyword || this.PeekToken(1).Kind != SyntaxKind.OpenParenToken))
                 {
-                    TypeSyntax type = this.ParseType(ParseTypeMode.AfterCase);
-                    if (!type.IsMissing)
+                    type = this.ParseType(ParseTypeMode.DefinitePattern);
+                    if (type.IsMissing)
                     {
-                        // X.Y.Z id
-                        if (this.IsTrueIdentifier() && this.CurrentToken.ContextualKind != SyntaxKind.WhenKeyword)
-                        {
-                            var designation = ParseSimpleDesignation();
-                            node = _syntaxFactory.DeclarationPattern(type, designation);
-                        }
+                        // either it is not shaped like a type, or it is a constant expression.
+                        this.Reset(ref resetPoint);
+                        type = null;
+                    }
+                }
+
+                PropertySubpatternSyntax propertySubpattern = null;
+                bool parsePropertySubpattern()
+                {
+                    if (this.CurrentToken.Kind == SyntaxKind.OpenBraceToken)
+                    {
+                        propertySubpattern = ParsePropertySubpattern();
+                        return true;
+                    }
+
+                    return false;
+                }
+
+                VariableDesignationSyntax designation = null;
+                bool parseDesignation()
+                {
+                    if (this.IsTrueIdentifier() && (!forCase || this.CurrentToken.ContextualKind != SyntaxKind.WhenKeyword))
+                    {
+                        designation = ParseSimpleDesignation();
```

Comment:
>designation [](start = 24, length = 11)

I think the code is much easier to follow (easier to track where things are changing) when this local is passed as an out parameter rather than referenced directly.

Reply:
Recorded in https://github.com/dotnet/roslyn/issues/23222.